File: apps/web-enterprise/src/lib/api/analytics.ts
// apps/web-enterprise/src/lib/api/analytics.ts
// Complete ITstep Analytics API client based on real DWH data

import { api } from "./config"

// ==================== TYPES ====================

export interface DateRange {
  start_date: string
  end_date: string
}

export interface MetricBase {
  impressions: number
  clicks: number
  spend: number
  conversions: number
  revenue: number
  ctr: number
  cpc: number
  cpm: number
  cvr: number
  cpa: number
  roas: number
}

export interface DashboardOverview {
  date_range: DateRange
  total_spend: number
  total_revenue: number
  total_conversions: number
  total_leads: number
  roas: number
  conversion_rate: number
  active_campaigns: number
  active_creatives: number
  spend_trend: number
  revenue_trend: number
  roas_trend: number
}

export interface RealTimeMetrics {
  active_sessions: number
  new_leads_today: number
  revenue_today: number
  conversions_today: number
  top_performing_creative: string | null
  alerts: string[]
  last_updated: string
}

export interface CampaignPerformance {
  campaign_id: string
  campaign_name: string
  platform: string
  total_metrics: MetricBase
  performance: {
    days_active: number
    first_seen: string | null
    last_active: string | null
    avg_cost_share: number
    avg_revenue_share: number
  }
}

export interface CreativePerformance {
  creative_id: string
  creative_name: string
  campaign_key?: string
  impressions: number
  clicks: number
  spend: number
  conversions: number
  revenue: number
  ctr: number
  cpc: number
  cpm: number
  roas: number
  cvr: number
  creative_url?: string
}

export interface CreativeBurnout {
  creative_id: string
  creative_name: string
  days_active: number
  initial_ctr: number
  current_ctr: number
  burnout_score: number
  status: 'fresh' | 'declining' | 'burned_out'
}

export interface FunnelStage {
  stage: string
  count: number
  conversion_rate: number
  drop_off_rate: number
}

export interface AnalyticsFilters {
  start_date: string
  end_date: string
  platforms?: string[]
  campaigns?: string[]
  products?: string[]
  branches?: number[]
  page?: number
  page_size?: number
  sort_by?: string
  sort_order?: 'asc' | 'desc'
}

// ==================== ANALYTICS API CLASS ====================

export class AnalyticsAPI {
  /**
   * Generic method for analytics requests
   */
  static async fetchAnalytics<T = any>(endpoint: string, params?: any): Promise<T> {
    try {
      const response = await api.get(endpoint, { params })
      return response.data
    } catch (error: any) {
      console.error("Analytics fetch error:", error)
      throw new Error(error.response?.data?.message || error.message || "Failed to fetch analytics")
    }
  }

  // ==================== DASHBOARD ====================

  /**
   * Get main dashboard overview
   */
  static async getDashboardOverview(dateRange: DateRange): Promise<DashboardOverview> {
    return this.fetchAnalytics("/analytics/overview/dashboard", dateRange)
  }

  /**
   * Get real-time metrics
   */
  static async getRealTimeMetrics(): Promise<RealTimeMetrics> {
    return this.fetchAnalytics("/analytics/overview/realtime")
  }

  /**
   * Get KPI metrics
   */
  static async getKPIs(dateRange: DateRange): Promise<any> {
    return this.fetchAnalytics("/analytics/overview/kpis", dateRange)
  }

  /**
   * Get platform performance
   */
  static async getPlatformPerformance(dateRange: DateRange): Promise<any[]> {
    return this.fetchAnalytics("/analytics/overview/platforms", dateRange)
  }

  // ==================== SALES ====================

  /**
   * Get sales revenue trend
   */
  static async getRevenueTrend(dateRange: DateRange): Promise<any> {
    return this.fetchAnalytics("/analytics/sales/revenue-trend", dateRange)
  }

  /**
   * Get sales by products
   */
  static async getSalesByProducts(dateRange: DateRange): Promise<any> {
    return this.fetchAnalytics("/analytics/sales/by-products", dateRange)
  }

  /**
   * Get conversion funnel
   */
  static async getConversionFunnel(dateRange: DateRange): Promise<any> {
    return this.fetchAnalytics("/analytics/sales/conversion-funnel", dateRange)
  }

  // ==================== CAMPAIGNS ====================

  /**
   * Get campaign performance
   */
  static async getCampaignPerformance(filters: AnalyticsFilters): Promise<{
    status: string
    data: CampaignPerformance[]
    total_count: number
  }> {
    return this.fetchAnalytics("/analytics/campaigns/performance", filters)
  }

  /**
   * Get daily trend for specific campaign
   */
  static async getCampaignDailyTrend(campaignId: string, dateRange: DateRange): Promise<any> {
    return this.fetchAnalytics(`/analytics/campaigns/daily-trend`, {
      campaign_id: campaignId,
      ...dateRange
    })
  }

  /**
   * Get campaigns grouped by products
   */
  static async getCampaignsByProducts(dateRange: DateRange): Promise<any> {
    return this.fetchAnalytics("/analytics/campaigns/by-products", dateRange)
  }

  /**
   * Get rolling performance metrics
   */
  static async getRollingPerformance(days: number = 7): Promise<any> {
    return this.fetchAnalytics("/analytics/campaigns/rolling-performance", { days })
  }

  /**
   * Get campaigns summary
   */
  static async getCampaignsSummary(dateRange: DateRange): Promise<any> {
    return this.fetchAnalytics("/analytics/campaigns/summary", dateRange)
  }

  /**
   * Get creatives for specific campaign
   */
  static async getCampaignCreatives(campaignId: string, dateRange: DateRange): Promise<any> {
    return this.fetchAnalytics(`/analytics/campaigns/${campaignId}/creatives`, dateRange)
  }

  // ==================== CREATIVES ====================

  /**
   * Get creative performance
   */
  static async getCreativePerformance(filters: AnalyticsFilters): Promise<{
    status: string
    data: CreativePerformance[]
    total_count: number
  }> {
    return this.fetchAnalytics("/analytics/creatives/performance", filters)
  }

  /**
   * Get creative burnout analysis
   */
  static async getCreativeBurnoutAnalysis(
    daysBack: number = 30,
    minDaysActive: number = 7
  ): Promise<{
    status: string
    data: CreativeBurnout[]
    alerts: string[]
  }> {
    return this.fetchAnalytics("/analytics/creatives/burnout-analysis", {
      days_back: daysBack,
      min_days_active: minDaysActive
    })
  }

  /**
   * Get top performing creatives
   */
  static async getTopPerformingCreatives(
    dateRange: DateRange,
    metric: 'roas' | 'revenue' | 'conversions' | 'ctr' | 'spend' = 'roas',
    limit: number = 10
  ): Promise<any> {
    return this.fetchAnalytics("/analytics/creatives/top-performers", {
      ...dateRange,
      metric,
      limit
    })
  }

  /**
   * Get creative themes analysis
   */
  static async getCreativeThemesAnalysis(dateRange: DateRange): Promise<any> {
    return this.fetchAnalytics("/analytics/creatives/themes-analysis", dateRange)
  }

  /**
   * Get detailed creative information
   */
  static async getCreativeDetails(creativeId: string, dateRange: DateRange): Promise<any> {
    return this.fetchAnalytics(`/analytics/creatives/${creativeId}/details`, dateRange)
  }

  // ==================== MARKETING ANALYTICS ====================

  /**
   * Get executive overview
   */
  static async getExecutiveOverview(dateRange?: DateRange, platform?: string): Promise<any> {
    const params: any = {}
    if (dateRange?.start_date) params.date_from = dateRange.start_date
    if (dateRange?.end_date) params.date_to = dateRange.end_date
    if (platform) params.platform = platform
    return this.fetchAnalytics("/analytics/marketing/executive-overview", params)
  }

  /**
   * Get channels and sources analysis
   */
  static async getChannelsSources(dateRange?: DateRange, platform?: string): Promise<any> {
    const params: any = {}
    if (dateRange?.start_date) params.date_from = dateRange.start_date
    if (dateRange?.end_date) params.date_to = dateRange.end_date
    if (platform) params.platform = platform
    return this.fetchAnalytics("/analytics/marketing/channels-sources", params)
  }

  /**
   * Get marketing campaigns data
   */
  static async getMarketingCampaigns(dateRange?: DateRange, platform?: string, campaignKey?: string): Promise<any> {
    const params: any = {}
    if (dateRange?.start_date) params.date_from = dateRange.start_date
    if (dateRange?.end_date) params.date_to = dateRange.end_date
    if (platform) params.platform = platform
    if (campaignKey) params.campaign_key = campaignKey
    return this.fetchAnalytics("/analytics/marketing/campaigns", params)
  }

  /**
   * Get marketing creatives data
   */
  static async getMarketingCreatives(params: {
    dateRange?: DateRange
    platform?: string
    searchText?: string
  }): Promise<any> {
    const queryParams: any = {}
    if (params.dateRange?.start_date) queryParams.date_from = params.dateRange.start_date
    if (params.dateRange?.end_date) queryParams.date_to = params.dateRange.end_date
    if (params.platform) queryParams.platform = params.platform
    if (params.searchText) queryParams.search_text = params.searchText
    return this.fetchAnalytics("/analytics/marketing/creatives", queryParams)
  }

  /**
   * Get CRM outcomes and 360 data
   */
  static async getCRMOutcomes(dateRange?: DateRange, platform?: string, source?: string): Promise<any> {
    const params: any = {}
    if (dateRange?.start_date) params.date_from = dateRange.start_date
    if (dateRange?.end_date) params.date_to = dateRange.end_date
    if (platform) params.platform = platform
    if (source) params.source = source
    return this.fetchAnalytics("/analytics/marketing/crm-outcomes", params)
  }

  /**
   * Get attribution funnel analysis
   */
  static async getAttributionFunnel(dateRange?: DateRange, platform?: string, productKey?: string): Promise<any> {
    const params: any = {}
    if (dateRange?.start_date) params.date_from = dateRange.start_date
    if (dateRange?.end_date) params.date_to = dateRange.end_date
    if (platform) params.platform = platform
    if (productKey) params.product_key = productKey
    return this.fetchAnalytics("/analytics/marketing/attribution-funnel", params)
  }

  /**
   * Get product performance analysis
   */
  static async getProductPerformance(dateRange?: DateRange, platform?: string): Promise<any> {
    const params: any = {}
    if (dateRange?.start_date) params.date_from = dateRange.start_date
    if (dateRange?.end_date) params.date_to = dateRange.end_date
    if (platform) params.platform = platform
    return this.fetchAnalytics("/analytics/marketing/product-performance", params)
  }

  /**
   * Get geography analysis
   */
  static async getGeographyAnalysis(dateRange?: DateRange, platform?: string): Promise<any> {
    const params: any = {}
    if (dateRange?.start_date) params.date_from = dateRange.start_date
    if (dateRange?.end_date) params.date_to = dateRange.end_date
    if (platform) params.platform = platform
    return this.fetchAnalytics("/analytics/marketing/geography-analysis", params)
  }

  /**
   * Get data quality metrics
   */
  static async getDataQuality(dateRange?: DateRange): Promise<any> {
    const params: any = {}
    if (dateRange?.start_date) params.date_from = dateRange.start_date
    if (dateRange?.end_date) params.date_to = dateRange.end_date
    return this.fetchAnalytics("/analytics/marketing/data-quality", params)
  }

  /**
   * Get available date range
   */
  static async getAvailableDateRange(): Promise<any> {
    return this.fetchAnalytics("/analytics/marketing/date-range")
  }

  // ==================== LEGACY SUPPORT ====================

  /**
   * Get sales analytics (legacy)
   */
  static async getSalesAnalytics(): Promise<any> {
    return this.fetchAnalytics("/analytics/sales/test")
  }

  /**
   * Get debug information
   */
  static async getDebugInfo(): Promise<any> {
    return this.fetchAnalytics("/analytics/sales/debug")
  }

  /**
   * Получить аналитику членства в организации
   */
  static async getMembershipStats(orgId: string): Promise<any> {
    return this.fetchAnalytics(`/orgs/${orgId}/memberships/stats`)
  }

  /**
   * Получить AI инсайты по продажам
   */
  static async getSalesInsights(): Promise<any> {
    return this.fetchAnalytics("/insights/sales")
  }
}

// Backward compatibility
export const fetchAnalytics = AnalyticsAPI.fetchAnalytics
export default AnalyticsAPI

File: apps/api/liderix_api/schemas/analytics.py
"""
Analytics Schemas - Complete schemas for ITstep Analytics Dashboard
Based on real ITstep database structure and requirements
"""
from datetime import datetime, date
from decimal import Decimal
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field
from enum import Enum


# ==================== ENUMS ====================

class PlatformType(str, Enum):
    FACEBOOK = "Facebook"
    GOOGLE = "Google"
    INSTAGRAM = "Instagram"
    YOUTUBE = "YouTube"
    OTHER = "Other"


class CampaignStatus(str, Enum):
    ACTIVE = "active"
    PAUSED = "paused"
    ENDED = "ended"


class CreativeType(str, Enum):
    IMAGE = "image"
    VIDEO = "video"
    CAROUSEL = "carousel"
    TEXT = "text"


class CreativeTheme(str, Enum):
    CAREER = "Кар'єра"
    MONEY = "Гроші"
    CHANGE = "Зміни"
    CERTIFICATES = "Сертифікати"
    GUARANTEES = "Гарантії"
    FROM_ZERO = "Навчання з нуля"
    EMOTIONAL = "Емоційна підтримка"


class LeadStatus(str, Enum):
    NEW = "new"
    CONTACTED = "contacted"
    QUALIFIED = "qualified"
    CONVERTED = "converted"
    LOST = "lost"


class Temperature(str, Enum):
    HOT = "hot"
    WARM = "warm"
    COLD = "cold"


# ==================== BASE SCHEMAS ====================

class DateRangeFilter(BaseModel):
    start_date: date
    end_date: date

    class Config:
        schema_extra = {
            "example": {
                "start_date": "2025-09-01",
                "end_date": "2025-09-30"
            }
        }


class MetricBase(BaseModel):
    """Base model for all metrics"""
    impressions: int = 0
    clicks: int = 0
    spend: Decimal = Field(default=0, decimal_places=6)
    conversions: int = 0
    revenue: Decimal = Field(default=0, decimal_places=2)

    @property
    def ctr(self) -> float:
        return (self.clicks / self.impressions * 100) if self.impressions > 0 else 0

    @property
    def cpc(self) -> float:
        return float(self.spend / self.clicks) if self.clicks > 0 else 0

    @property
    def cpm(self) -> float:
        return float(self.spend / self.impressions * 1000) if self.impressions > 0 else 0

    @property
    def cvr(self) -> float:
        return (self.conversions / self.clicks * 100) if self.clicks > 0 else 0

    @property
    def cpa(self) -> float:
        return float(self.spend / self.conversions) if self.conversions > 0 else 0

    @property
    def roas(self) -> float:
        return float(self.revenue / self.spend) if self.spend > 0 else 0


# ==================== CAMPAIGNS ====================

class Campaign(BaseModel):
    """Campaign information"""
    id: str
    name: str
    platform: PlatformType
    product_key: Optional[str] = None
    branch_id: Optional[int] = None
    status: CampaignStatus = CampaignStatus.ACTIVE
    first_seen: Optional[date] = None
    last_active_date: Optional[date] = None


class CampaignMetrics(MetricBase):
    """Campaign daily metrics"""
    campaign_id: str
    campaign_name: str
    date: date
    platform: PlatformType
    reach: Optional[int] = None
    frequency: Optional[float] = None

    # Share metrics
    share_cost_in_platform: Optional[float] = None
    share_revenue_in_platform: Optional[float] = None
    share_contracts_in_platform: Optional[float] = None


# ==================== CREATIVES ====================

class Creative(BaseModel):
    """Creative information"""
    id: str
    name: str
    campaign_id: Optional[str] = None
    type: CreativeType = CreativeType.IMAGE
    platform: PlatformType = PlatformType.FACEBOOK
    title: Optional[str] = None
    theme: Optional[CreativeTheme] = None
    is_personalized: bool = False
    fb_creative_id: Optional[str] = None
    creative_url: Optional[str] = None
    status: CampaignStatus = CampaignStatus.ACTIVE


class CreativeMetrics(MetricBase):
    """Creative daily metrics"""
    creative_id: str
    creative_name: str
    date: date
    campaign_key: Optional[str] = None
    reach: Optional[int] = None
    frequency: Optional[float] = None

    # Allocation metrics
    cost_alloc_by_revenue: Optional[Decimal] = None
    cost_alloc_by_contracts: Optional[Decimal] = None


class CreativeBurnout(BaseModel):
    """Creative burnout analysis"""
    creative_id: str
    creative_name: str
    days_active: int
    initial_ctr: float
    current_ctr: float
    burnout_score: float  # 0-100, higher = more burned out
    status: str  # "fresh", "declining", "burned_out"

    @property
    def needs_refresh(self) -> bool:
        return self.burnout_score > 70 or (self.days_active > 14 and self.current_ctr < self.initial_ctr * 0.5)


# ==================== DASHBOARD RESPONSES ====================

class DashboardOverview(BaseModel):
    """Main dashboard overview"""
    date_range: DateRangeFilter
    total_spend: Decimal
    total_revenue: Decimal
    total_conversions: int
    total_leads: int
    roas: float
    conversion_rate: float
    active_campaigns: int
    active_creatives: int

    # Trends (compared to previous period)
    spend_trend: float  # % change
    revenue_trend: float
    roas_trend: float


class PlatformPerformance(BaseModel):
    """Platform performance summary"""
    platform: PlatformType
    campaigns_count: int
    creatives_count: int
    metrics: MetricBase
    share_of_spend: float
    share_of_revenue: float
    performance_score: float  # 0-100


class FunnelStage(BaseModel):
    """Funnel stage data"""
    stage: str
    count: int
    conversion_rate: float
    avg_time_in_stage: Optional[float] = None  # hours
    drop_off_rate: float


class FunnelAnalysis(BaseModel):
    """Complete funnel analysis"""
    date_range: DateRangeFilter
    stages: List[FunnelStage]
    overall_conversion_rate: float
    total_drop_offs: int
    bottleneck_stage: str


class RealTimeMetrics(BaseModel):
    """Real-time metrics for dashboard"""
    active_sessions: int
    new_leads_today: int
    revenue_today: Decimal
    conversions_today: int
    top_performing_creative: Optional[str] = None
    alerts: List[str] = []
    last_updated: datetime


class AnalyticsFilters(BaseModel):
    """Common filters for analytics endpoints"""
    date_range: DateRangeFilter
    platforms: Optional[List[PlatformType]] = None
    campaigns: Optional[List[str]] = None
    products: Optional[List[str]] = None
    branches: Optional[List[int]] = None

    # Pagination
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=50, ge=1, le=200)

    # Sorting
    sort_by: str = "spend"
    sort_order: str = Field(default="desc", pattern="^(asc|desc)$")


class TimeSeriesPoint(BaseModel):
    """Time series data point"""
    date: date
    value: float
    metric: str


# ==================== LEGACY COMPATIBILITY ====================

class KPIMetrics(BaseModel):
    revenue: float
    profit: float
    cr: float   # Conversion Rate
    cac: float  # Customer Acquisition Cost


class RevenueItem(BaseModel):
    date: date
    revenue: float


class AdsPerformanceItem(BaseModel):
    report_date: date
    campaign_id: str
    impressions: float
    clicks: float
    conversions: float
    cost: float
    cpc: Optional[float] = None
    cpa: Optional[float] = None
    ctr: Optional[float] = None


class TopCreativeItem(BaseModel):
    date: date
    creative_id: str
    campaign_id: str
    impressions: float
    clicks: float
    spend: float
    ctr: float
    cpc: float


class DeviceItem(BaseModel):
    device_category: str
    sessions: int
    users: int
    conversions: int
    conversion_rate: float

# 👤 Клиентская метрика
class CustomerMetricsItem(BaseModel):
    report_date: date
    new_customers: int
    returning_customers: int
    avg_check: float
    clv: float

# 📊 ROAS по кампаниям
class CampaignROASItem(BaseModel):
    campaign_id: str
    campaign_name: str
    spend: float
    revenue: float
    roas: float

# 🌐 Трафик по каналам
class TrafficItem(BaseModel):
    date: date
    channel: str
    total_sessions: int
    total_users: int
    new_users: int
    engaged_sessions: int
    avg_session_duration: float
    avg_bounce_rate: float
    avg_engagement_rate: float

# 🔍 AI-инсайты
class AIInsight(BaseModel):
    summary: str
    insights: List[str]
    recommendations: List[str]

# 🔮 Прогноз
class ForecastItem(BaseModel):
    metric: str
    value: float
    predicted_value: float
    delta: float

# 🧩 Воронка
class FunnelStep(BaseModel):
    step_name: str
    users: int
    conversion_rate: float

# 📈 Удержание
class RetentionCohort(BaseModel):
    cohort_week: str
    week_0: int
    week_1: Optional[int] = None
    week_2: Optional[int] = None
    week_3: Optional[int] = None
    week_4: Optional[int] = None